File: src/c#/FuzzySearch/FuzzySearchService.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace FuzzySearch
{
    public class FuzzySearchService
    {
        private Dictionary<string, List<string>> index;

        #region Soundex
        private string EncodeChar(char c)
        {
            switch (Char.ToLower(c))
            {
                case 'b':
                case 'f':
                case 'p':
                case 'v':
                    return "1";
                case 'c':
                case 'g':
                case 'j':
                case 'k':
                case 'q':
                case 's':
                case 'x':
                case 'z':
                    return "2";
                case 'd':
                case 't':
                    return "3";
                case 'l':
                    return "4";
                case 'm':
                case 'n':
                    return "5";
                case 'r':
                    return "6";
                default:
                    return string.Empty;
            }
        }

        private string Soundex(string data)
        {
            StringBuilder result = new StringBuilder();

            if (!string.IsNullOrEmpty(data))
            {
                string previousCode, currentCode;
                result.Append(Char.ToUpper(data[0]));
                previousCode = string.Empty;

                for (int i = 1; i < data.Length; i++)
                {
                    currentCode = EncodeChar(data[i]);

                    if (currentCode != previousCode)
                        result.Append(currentCode);

                    if (result.Length == 4) break;

                    if (!currentCode.Equals(string.Empty))
                        previousCode = currentCode;
                }
            }
            if (result.Length < 4)
                result.Append(new String('0', 4 - result.Length));

            return result.ToString();
        }
        #endregion

        private IEnumerable<string> ParseValue(string value)
        {
            return value.ToLower().Split(' ');
        }

        private Dictionary<string, List<string>> Init(IEnumerable<string> valuesToIndex)
        {
            var indexToBuild = new Dictionary<string, List<string>>();

            foreach (var item in valuesToIndex)
            {
                var parsedInputs = ParseValue(item);
                foreach (var value in parsedInputs)
                {
                    var hash = Soundex(value);
                    List<string> currentlyIndexedItems;
                    if (indexToBuild.TryGetValue(hash, out currentlyIndexedItems))
                    {
                        currentlyIndexedItems.Add(item);
                    }
                    else
                    {
                        indexToBuild.Add(hash, new List<string> { item });
                    }
                }
            }
            return indexToBuild;
        }

        public FuzzySearchService(IEnumerable<string> valuesToIndex)
        {
            index = Init(valuesToIndex);
        }

        public IEnumerable<Result> Search(string inputToSearch)
        {
            List<Result> results = new List<Result>();
            foreach (var input in ParseValue(inputToSearch))
            {
                List<string> valuesFromIndex;
                if(index.TryGetValue(Soundex(input), out valuesFromIndex))
                {
                    foreach (var resultFromIndex in valuesFromIndex)
                    {
                        results.Add(new Result(inputToSearch, resultFromIndex, Score(inputToSearch, resultFromIndex)));
                    }
                }
            }

            return results;
        }

        private double Score(string in1, string in2)
        {
            return DicesCoeffienct(in1.ToLower(), in2.ToLower())*100;
        }

        private double DicesCoeffienct(string in1, string in2)
        {
            HashSet<string> nx = new HashSet<string>();
            HashSet<string> ny = new HashSet<string>();
 
            for(int i = 0; i < in1.Length - 1; i++)
            {
                char x1 = in1[i];
                char x2 = in1[i + 1];
                string temp = x1.ToString() + x2.ToString();
                nx.Add(temp);
            }
            for(int j = 0; j < in2.Length - 1; j++)
            {
                char y1 = in2[j];
                char y2 = in2[j + 1];
                string temp = y1.ToString() + y2.ToString();
                ny.Add(temp);
            }
 
            HashSet<string> intersection = new HashSet<string>(nx);
            intersection.IntersectWith(ny);
 
            double dbOne = intersection.Count;
            return (2 * dbOne) / (nx.Count + ny.Count);
        }
    }
}

File: src/scala/src/fuzzysearch/FuzzySearchService.scala
package fuzzysearch

import collection.mutable.{ListBuffer, Map, HashSet}

class FuzzySearchService(wordsToIndex: List[String]){
  require(wordsToIndex != null)

  private val index: Map[String, List[String]] = this.init(wordsToIndex)

  private def encodeChar(value: Char) : Char = {
    value match{
      case 'b' => '1'
      case 'f' => '1'
      case 'p' => '1'
      case 'v' => '1'
      case 'c' => '2'
      case 'g' => '2'
      case 'j' => '2'
      case 'k' => '2'
      case 'q' => '2'
      case 's' => '2'
      case 'x' => '2'
      case 'z' => '2'
      case 'd' => '3'
      case 't' => '3'
      case 'l' => '4'
      case 'm' => '5'
      case 'n' => '5'
      case 'r' => '6'
      case _ => 0
    }
  }

  //TODO: This can probably be done better in an inline fashion
  private def padSoundex(value: StringBuilder) : StringBuilder = {
    var result : StringBuilder = value
    if (result.length != 4)
    {
      if (result.length == 1) result = result + '0' + '0' + '0'
      else if (result.length == 2) result = result + '0' + '0'
      else if (result.length == 3) result = result + '0'
    }
    result
  }

  private def soundex(data: String) : String = {
    var result : StringBuilder = new StringBuilder(4)

    if (data != null && data.length() > 0) {
      var previousCode : Char = Char.MinValue
      var currentCode : Char = Char.MinValue

      val charArray : Array[Char] = data.toCharArray

      result +=  data.charAt(0).toUpper

      for (index <- 1 to charArray.length - 1){
        val encodedChar: Char = encodeChar(charArray(index).toLower)
        if (encodedChar != 0)
          currentCode = encodedChar

        if (currentCode != previousCode)
          result += currentCode

        if (result.length == 4)
          return result.toString()

        if (currentCode != 0)
          previousCode = currentCode
      }
    }
    padSoundex(result).toString()
  }

  private def init(phrasesToIndex: List[String]) : Map[String, List[String]] = {
    val indexToBuild = Map[String, List[String]]()

    for (phrase <- phrasesToIndex){
      val words = phrase.split(" ")
      for (word <- words){
        val hash = soundex(word)
        if (indexToBuild.contains(hash)){
          //val value: List[String] = indexToBuild.get(hash).get
          indexToBuild += hash -> (phrase :: indexToBuild.get(hash).get).toList
        }else {
          indexToBuild += hash -> List(phrase)
        }
      }
    }
    //println(indexToBuild)
    indexToBuild
  }

  def search(input: String) : List[Result] = {
    val result: ListBuffer[Result] = ListBuffer()

    for (valuesToSearch <- input.split(" ")){
      val valuesFromIndex = index.get(soundex(valuesToSearch))
      if(valuesFromIndex != None){
        val results = valuesFromIndex.get.toSet
        for (value <- results){
          result.append(new Result(input, value, score(input, value)))
        }
      }
    }
    result.toList
  }

  private def score(input1: String, input2: String) : Double = {
    diceCoefficient(input1.toLowerCase, input2.toLowerCase) * 100
  }

  private def diceCoefficient(input1: String, input2: String) : Double = {
    val set1: HashSet[String] = HashSet()
    val set2: HashSet[String] = HashSet()

    for(index <- 0 to input1.length - 2){
      val x1: Char = input1(index)
      val x2: Char = input1(index + 1)
      val tmp: String = new String(Array(x1, x2))
      set1.add(tmp)
    }
    for(index <- 0 to input2.length - 2){
      val x1: Char = input2(index)
      val x2: Char = input2(index + 1)
      val tmp: String = new String(Array(x1, x2))
      set2.add(tmp)
    }

    val intersection : HashSet[String] = set1.intersect(set2)
    (intersection.size * 2.0)/(set1.size + set2.size)
  }

}

case class Result(search: String, result: String, score: Double) {
  override def toString() = {
    "Searched: " + search + " Result: " + result + " Score: " + score.toString
  }
}